### ISSUE TITLE
[Feature] Support Custom Types within data inspector

### ISSUE THREAD

**Issue by m10653** [WerWolv/ImHex#1849]
### What feature would you like to see?

It would be fantastic to allow custom types from the pattern editor to be used/added within the data inspector tab. This would allow for simple processing of highlighted data and quick verification of custom value types. Even if it is just basic value support/display without the support of viewing nested structures it still would be very useful. 

### How will this feature be useful to you and others?

This would allow for the quick evaluation of custom data types on bytes in a file without needing to type into the pattern editor. Making it much easer to identify valid values in unknown file formats.

### Request Type

- [ ] I can provide a PoC for this feature or am willing to work on it myself and submit a PR

**Comment by paxcut** [comment]
If the issue is the rapid evaluation of patterns without typing you already can do that:
1) Right click on the byte you want to place a pattern and 
2) select `PlacePattern > Custom` to see a list of patterns defined in the pattern editor. 
3) Choose one and the pattern will show on the pattern data window the same way it would show if you typed the placement and ran it.

For the data inspector there is also the possibility to define custom entries using the Pattern Language. For details see the [docs](https://docs.werwolv.net/imhex/views/data-inspector#adding-new-rows)

**Comment by m10653** [comment]
So the custom entries approach with the pattern lang is exactly what I am looking for however in a way that is much easer to access. It would be nice when a user hits edit within the data inspector to enable or disable inspectors to click "add" and that would allow the user to select a dropdown of custom types that are within the pattern editor. IE the same behavior as the place pattern custom menu. 

Right now its quite cumbersome if you want to click around a file and see the values of a custom type and see if they are valid or not.

**Comment by paxcut** [comment]
I don't agree that adding a context menu to the data inspector is that much different than what you call quite cumbersome. You still would need to click on the data for every place to be checked. 1 click + 1 menu per entry  versus 1 click +1 menu + 1 click per entry does not change quite cumbersome to not quite cumbersome. The real way to make it nice is to avoid menus that need to be clicked and there could be an easy way to use any pattern defined in the files being edited in the pattern editor.

All that needs to be implemented is a way to be able to import the current pattern from the file where the `CustomTypeInspector` function is located. If Imhex can get a list of all the patterns to create the placement menu, then it has to be able to get the file where they are defined which should be much easier, so invent an alias for the current active pattern file in the editor that when it is used to import it imports the contents of the pattern editor. I have a feeling that the virtual file system could be use for this purpose but it isn't documented so I don't know what it can or can't do.

If that was possible then simply define `using CustomType = SomeStructInYourPattern` in your pattern file and copy the following custom inspector to the `scripts/inspectors` folder.
```cpp  
import std.io;
import CurrentActivePattern;

struct CustomTypeInspector {
    CustomType value;
} [[format_read("custom_type_inspector_formatter")]];

fn custom_type_inspector_formatter(auto customTypeInspector) {
    return std::format("{}", customTypeInspector.value);
};

CustomTypeInspector customTypeInspector @ $ [[name("Custom Type")]];
```
To change the pattern being checked change the value of `CustomType` being assigned to to the new type.  
 You can also use the newly create `IMPORTED` macro to avoid placing variables when the pattern is imported. 
 This method is much easier to implement and cleaner than creating menu entries and is also much more powerful since you can create as many entries as you want and you can also configure what will be displayed in the data inspector and how it will be displayed which the menu does not. If you dont care use {}.

**Comment by m10653** [comment]
I think we are talking about the same thing, or at least close to it.

The workflow would be as follows.
1. Develop/work on pattern pattern editor that has custom imported or defined types.
2. Want to be able to preview the parsing of these types based on their cursor location within the hex editor and for the results to be displayed within the data inspector.
3. Click the data inspector edit button + ( New Add button that lets the user select from types available within the pattern editor)
4. New field is added to the data inspector for that type and its value updates based on the cursors selected position just like the other types.

Right now it seems like there are two ways to get a similar behavior using the scripts/inspectors folder but that requires leaving the editor and does not update real time with changes in the pattern editor.  Or right clicking and placing pattern. The placing pattern does not really solve the problem as it places the value in the pattern data view and would require you to right click and add pattern for every byte you wish to check. (or updating the byte offset in the pattern file)


Your proposal of having the ability to import the current pattern and add it does seem to work although I feel that might bit a bit cluttered when importing external pattern files where listing all the types in the data inspector would not be desired. I am aware that you can disable the display within the inspector but displaying all of them by default seems a tad problematic. 
Example of lots of imported/nested types that might not be great to import
![image](https://github.com/user-attachments/assets/ee9afd59-655f-4196-a389-9ea27a390693)

**Comment by paxcut** [comment]
> Right now it seems like there are two ways to get a similar behavior using the scripts/inspectors folder but that requires leaving the editor and does not update real time with changes in the pattern editor.

The custom inspector option is a one time edit. Once the function is in place you don't have to leave the editor to update the patterns that will be used to sample the data. You will need to edit the definition of the type used but that will be done in your pattern. It will also save you from having to click on the data inspector to chose the pattern.  

Using menus to chose patterns can go bad very quickly when you use complex patterns as you clearly show. It is best to avoid that completely. It is always going to be better to simply change the name of the pattern that is chosen for the inspector. For example if you wanted to use `DosTimeConverter` as the pattern for the inspector you would have to navigate that same set of menus instead of just typing the name in the pattern editor.
 
> Your proposal of having the ability to import the current pattern and add it does seem to work although I feel that might bit a bit cluttered when importing external pattern files where listing all the types in the data inspector would not be desired.

The proposal I made didn't include listing all the types in the data inspector or at least it didn't mean to.Importing the pattern makes all of them available but you would only display one or a few of them. Also you wouldn't use menus to access them so the example of the long chain of types does not apply. Maybe my description of the proposal was confusing if you drew those conclusions from it. 

You say the proposal seems to work but there is no way to import the pattern editor currently loaded file so that if you change to a different one you then would have the option to display the patterns in the new file without having to edit the inspector function. However that is the only feature that would need implementation and adding that has to be much easier than adding a menu on the inspector that lists all the patterns in the file.
